File: SYSC 2100A/Lab2/Fraction.py
# SYSC 2100 Winter 2021 Lab 2
#Ngo Huu Gia Bao-101163137


def gcd(m: int, n: int) -> int:
    """Return the greatest common divisor of m and n.

    Precondition: m > 0, n > 0.

    >>> gcd(4, 1)
    1
    >>> gcd(30, 5)
    5
    >>> gcd(9, 17)
    1
    """
    while m % n != 0:
        m, n = n, m % n
    return n


class Fraction:
    def __init__(self, top: int, bottom: int) -> None:
        """Initialize self with numerator top and denominator bottom.

        Raise a ValueError exception if bottom is 0.

        When __init__ returns, this fraction will be in reduced form.
        This means that:
        (1) if the numerator is equal to 0, the denominator is always 1.
        (2) if the numerator is not equal to 0, the denominator is always
        positive. This means that negative fractions always have a negative
        numerator and a positive denominator.
        (3) the numerator and denominator have no common divisors
        other than 1.

        >>> Fraction(3, 4)
        Fraction(3, 4)
        >>> Fraction(6, 8)
        Fraction(3, 4)
        >>> Fraction(-3, 4)
        Fraction(-3, 4)
        >>> Fraction(3, -4)
        Fraction(-3, 4)
        >>> Fraction(-3, -4)
        Fraction(3, 4)
        >>> Fraction(0, 5)
        Fraction(0, 1)
        """

        # This method must be modified to match the specification
        # in the docstring.
        
        commonDiv = gcd(top, bottom)
        self._num = top // commonDiv            
        self._den = bottom // commonDiv     

    def __str__(self) -> str:
        """Return a string representation of self in the format:
        top/bottom.

        >>> f = Fraction(3, 4)
        >>> str(f)
        '3/4'
        """
        return "{0}/{1}".format(self._num, self._den)

    def __repr__(self) -> str:
        """Return a string representation of self in the format:
        Fraction(top, bottom).

        >>> f = Fraction(3, 4)
        >>> f
        Fraction(3, 4)
        >>> f = Fraction(3, -4)
        >>> f
        Fraction(-3, 4)
        """
        return "Fraction({0},{1})\n".format(self._num,self._den)

    def numerator(self) -> int:
        """Return the numerator of self.

        >>> f = Fraction(3, 4)
        >>> f.numerator()
        3
        """
        return self._num

    def denominator(self) -> int:
        """Return the denominator of self.

        >>> f = Fraction(3, 4)
        >>> f.denominator()
        4
        """
        return self._den

    def __eq__(self, other_fraction) -> bool:
        """Return True if self equals other_fraction.

        >>> f1 = Fraction(3, 4)
        >>> f2 = Fraction(6, 8)
        >>> f1 == f2
        True
        >>> f2 = Fraction(1, 2)
        >>> f1 == f2
        False
        """
        first_num = self._num * other_fraction._den
        second_num = other_fraction._num * self._den

        return first_num == second_num

    def __lt__(self, other_fraction) -> bool:
        """Return True if self is less than other_fraction.

        >>> f1 = Fraction(3, 4)
        >>> f2 = Fraction(6, 8)
        >>> f1 < f2
        False
        >>> f2 = Fraction(7, 8)
        >>> f1 < f2
        True
        """
        first_num = self._num / self._den
        second_num = other_fraction._num / other_fraction._den
        
        return first_num < second_num
        

    def __le__(self, other_fraction) -> bool:
        """Return True if self is less than or equal to other_fraction.

        >>> f1 = Fraction(3, 4)
        >>> f2 = Fraction(6, 8)
        >>> f1 <= f2
        True
        >>> f2 = Fraction(7, 8)
        >>> f1 <= f2
        True
        >>> f2 <= f1
        False
        """
        first_num = self._num / self._den
        second_num = other_fraction._num / other_fraction._den
        
        return first_num <= second_num

    def __add__(self, other_fraction):
        """Return the sum of self and other_fraction.

        >>> f1 = Fraction(3, 4)
        >>> f2 = Fraction(1, 8)
        >>> f1 + f2
        Fraction(7, 8)
        >>> f1 = Fraction(1, 4)
        >>> f2 = Fraction(1, 2)
        >>> f1 + f2
        Fraction(3, 4)
        """
        new_num = (self._num * other_fraction._den
                   + self._den * other_fraction._num)
        new_den = self._den * other_fraction._den
        return Fraction(new_num, new_den)

    def __sub__(self, other_fraction):
        """Return the difference of self and other_fraction; that is,
        the Fraction that is obtained by subtracting other_fraction from sub.

        >>> f1 = Fraction(3, 4)
        >>> f2 = Fraction(1, 8)
        >>> f1 - f2
        Fraction(5, 8)
        >>> f1 = Fraction(1, 4)
        >>> f2 = Fraction(1, 2)
        >>> f1 - f2
        Fraction(-1, 4)
        """
        new_num = (self._num * other_fraction._den
                   - self._den * other_fraction._num)
        new_den = self._den * other_fraction._den
        return Fraction(new_num , new_den)
    

    def __mul__(self, other_fraction):
        """Return the product of self and other_fraction; that is,
        self - other_fraction.

        >>> f1 = Fraction(3, 4)
        >>> f2 = Fraction(1, 8)
        >>> f1 * f2
        Fraction(3, 32)
        >>> f2 = Fraction(6, 8)
        >>> f1 * f2
        Fraction(9, 16)
        """
        new_num = self._num * other_fraction._num
        new_den = self._den * other_fraction._den
        return Fraction(new_num, new_den) 

    def __truediv__(self, other_fraction):
        """Return the result of dividing self by other_fraction.

        >>> f1 = Fraction(3, 4)
        >>> f2 = Fraction(1, 8)
        >>> f1 / f2
        Fraction(6, 1)
        >>> f2 = Fraction(2, 1)
        >>> f1 * f2
        Fraction(3, 8)
        """
        new_num = self._num * other_fraction._den
        new_den = self._den * other_fraction._num
        return Fraction(new_num, new_den) 

if __name__ == "__main__":
    x = Fraction(1, 2)
    print('x = {0}'.format(x))
    y = Fraction(2, 3)
    print('y = {0}'.format(y))
    print('x + y = {0}'.format(x + y))
    print('x == y is {0}'.format(x == y))
    
    print ("\nEx 10:")
    a = Fraction(3,4)
    print("a = {0}".format(a))
    
    b = Fraction(4,5)
    print("b = {0}".format(b))
    print("a != b is {0}".format(a!=b))
    print("a <= b is {0}".format(a<=b))
    print("a >= b is {0}".format(a>=b))
    print("a > b is {0}".format(a>b))
    print("a < b is {0}".format(a<b))

File: SYSC 2006C/lab5/fraction_pointer/fraction.c
/* fraction.c - SYSC 2006 Lab 5 */

#include <stdlib.h>  // abs(x)
#include <stdio.h>   // printf
#include <assert.h>  // assert

#include "fraction.h"

/* Print the fraction pointed to by pf in the form a/b. 
 */
void print_fraction(const fraction_t *pf)
{
	printf("%d/%d", pf -> num, pf -> den);
}

/* Return the greatest common divisor of integers a and b; 
   i.e., return the largest positive integer that evenly divides 
   both values.
*/
int gcd(int a, int b)
{
	/* Euclid's algorithm, using iteration and calculation of remainders. */
	int q = abs(a);
	int p = abs(b);
	int r = q % p;

	while (r!=0) {
		q = p;
		p = r;
		r = q % p;

	}

	return p;
}

/* Convert the fraction pointed to by pf to reduced form.

   This means that:
   (1) if the numerator is equal to 0 the denominator is always 1.
   (2) if the numerator is not equal to 0 the denominator is always
       positive, and the numerator and denominator have no common
       divisors other than 1.

   In other words,
   (1) if the numerator is 0 the denominator is always changed to 1.
   (2) if the numerator and denominator are both negative, both values
       are made positive, or if the numerator is positive and the 
       denominator is negative, the numerator is made negative and the 
       denominator is made positive.
   (3) the numerator and denominator are both divided by their greatest
       common divisor. 
*/
void reduce(fraction_t *pf)
{
	int commonDiv = gcd (pf->num, pf->den);

	if (pf->num == 0) {
		pf->den = 1;
	}

	else if (pf->num > 0 && pf->den > 0) {
		pf->num = pf->num / commonDiv;
 		pf->den = pf->den / commonDiv;
	}

	else if (pf->num < 0 && pf->den <0) {
		pf->num = - pf->num / commonDiv;
 		pf->den = - pf->den / commonDiv;
	}

		
	else if (pf->num > 0 && pf->den < 0) {
		pf->num = - pf->num / commonDiv;
 		pf->den = - pf->den / commonDiv;
	}
	
	else if (pf->num < 0 && pf->den > 0) {

		pf->num = pf->num / commonDiv;
 		pf->den = pf->den / commonDiv;
	}

}

/* Initialize the fraction pointed to by new_fraction 
   with numerator a and denominator b.
   Print an error message and terminate the calling program via exit()
   if the denominator is 0.
   The fraction returned by this function is always in reduced form
   (see documentation for reduce).
*/
void make_fraction(int a, int b, fraction_t *new_fraction)
{
	new_fraction->num = a;
	new_fraction->den = b;

	if (b == 0) {
		printf ("Error, denominator cannot be 0");
		exit(0);
	} 
	else {
		reduce (new_fraction);
	}

}

/* Initialize the fraction pointed to by sum with the sum of 
   the fractions pointed to by pf1 and pf2.
   The fraction created by this function is always in reduced form
   (see documentation for reduce).
*/
void add_fractions(const fraction_t *pf1, const fraction_t *pf2, fraction_t *sum)
{
	sum -> num = pf1->num * pf2->den + pf2->num * pf1->den;
	sum -> den = pf1->den * pf2->den;

	make_fraction(sum -> num, sum -> den, sum);

}

/* Initialize the fraction pointed to by product with the product of 
   the fractions pointed to by pf1 and pf2.
   The fraction created by this function is always in reduced form
   (see documentation for reduce).
*/
void multiply_fractions(const fraction_t *pf1, const fraction_t *pf2, 
	                    fraction_t *product)
{
	product->num = pf1->num * pf2->num;
	product->den = pf1->den * pf2->den;
	
	make_fraction(product->num, product->den, product);
}
